### ISSUE TITLE
Reduce scope of the "Convert to throw expression" refactoring

### ISSUE THREAD

**Issue by hughbe** [dotnet/roslyn#20910]
I recently discovered that the unit tests for one of my projects had increased in length by several seconds, and found the "Use throw exception was responsible (I recently did a batch conversion for all my sins.)

Visual Studio collapsed the `if (type == null)` check before the call to `new ColumnReader`, which is an expensive operation, to after the expensive call.

Specifically,
```cs
if (type == null)
{
    throw new ArgumentNullException(nameof(type));
}

Reader = new ColumnReader(columnName);
Type = type;
```

was transformed to
```cs
Reader = new ColumnReader(columnName);
Type = type ?? throw new ArgumentNullException(nameof(columnName));
```

Not only can this cause unwanted performance effects that aren't obvious.

This can also cause breaking changes to APIs. This is something I found using Visual Studio writing contributions to dotnet/corefx. Visual Studio suggests using a throw expression, and this changes the order of validation of arguments in the implementation.
In the example above, ColumnReader may throw an exception validating `columnName`, whereas it did not validate `columnName` before. This is a breaking change.

I suggest that the refactoring should not be offered if there is a method, constructor or property call between the old throw block and the property assignment.

The full code sample is below. I've added a `Thread.Sleep` statement to simulate a call to the database that caused this issue previously:
```cs
using System;
using System.Threading;

namespace Test
{
    public class Program
    {
        public static void Main()
        {
        }
    }

    public class ColumnMapper
    {
        public ColumnReader Reader { get; }
        public Type Type { get; }

        public ColumnMapper(Type type, string columnName)
        {
            // Transforming this code changes argument validation and can cause unwanted performance
            // side-effects
            if (type == null)
            {
                throw new ArgumentNullException(nameof(type));
            }

            Reader = new ColumnReader(columnName);
            Type = type;
        }
    }

    public class ColumnReader
    {
        public ColumnReader(string columnName)
        {
            // Some expensive operation.
            Thread.Sleep(2000);
        }
    }
}
```

**Comment by Pilchie** [comment]
@CyrusNajmabadi did we already make this change to not offer this suggestion if there is code between the check and the assignment for 15.3?

**Comment by Pilchie** [comment]
(Note - we do not appear to have changed it).

**Comment by CyrusNajmabadi** [comment]
No.  We have currently accepted that this is just a suggestion and doesn't necessarily completely preserve semantics.  The core issue is that if we want to preserve semantics, we effectively will not ever be able to offer this change.  That's because even the simple case of two-params+two-checks+two-assignemnts isn't in the form:

check/assign/check/assign.  It's:

check/check/assign/assign.  So the checks always have intervening code.  At the time we accepted that small changes in behavior might be possible.

One thing we did consider was making it clearer that this action could change semantics.  For example, we could have a warning glyph shown with the code action and a message in the preview that says that semantics may change.
